### PR TITLE
GOV-241: Batch files go through a final check to be clean of defects and inconsistencies

### PR DESCRIPTION
**Pull request by ankita10r**
## Description

Contains batch file validations
## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Design related bullet points or design document link related to this PR added in the description above. 

- [ ] Updated corresponding Postman Collection or Api documentation for the changes in this PR.

- [x] Create/update unit or integration tests for verifying the changes made.

- [ ] Add required Swagger annotation and update API documentation with details of any API changes if applicable

- [x] Followed the naming conventions as given in https://docs.google.com/document/d/1Q4vaMSzrTxxh9TS0RILuNkSkYCxotuYk1Xe0CMIkkCU/edit?usp=sharing

### CODE REVIEW

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java [hunk 1]
```diff
@@ -233,19 +267,158 @@ private void setup() {
...
+        InputStream ips = new FileInputStream(file);
+        InputStreamReader ipsr = new InputStreamReader(ips);
+        BufferedReader br = new BufferedReader(ipsr);
+        String line;
+        br.readLine();
+        while ((line = br.readLine()) != null) {
+            String[] row = line.split(",");
+            logger.info("row is" + Arrays.toString(row));
+            if (row.length != columnNames.size()) {
+                logger.info("Row invalid " + row.length + " " + columnNames.size());
```
Comment (fynmanoj) [inline src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:341]: we have to provide more informative error messages,  please add a ticket to the back log to aggregate these error messages

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java [hunk 2]
```diff
@@ -233,19 +267,158 @@ private void setup() {
...
+            logger.info("Columns in the csv file are " + Arrays.toString(columns));
+        }
+        int i = 0;
+        while(i < columns.length) {
+            if(columnNames.contains(columns[i])) {
+                logger.info("Column name " + columns[i] + " is at index " + columnNames.indexOf(columns[i]));
+                i++;
+
+            } else {
+                logger.info("Column name " + columns[i] + " is at index " + columnNames.indexOf(columns[i]));
```
Comment (fynmanoj) [inline src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:398]: when contains operation is false, how are you expecting this `columnNames.indexOf(columns[i]` to work?

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java [hunk 3]
```diff
@@ -133,17 +166,16 @@ private void setup() {
                     logger.info("\n\n Filename: " + fileName + " \n\n");
```
Comment (fynmanoj) [inline src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:166]: do not use + operator to concatenate logs , use '{}' format

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java [hunk 4]
```diff
@@ -233,19 +267,158 @@ private void setup() {
...
+                .process(exchange -> {
+                    System.setProperty("mail.mime.multipart.ignoreexistingboundaryparameter", "true");
+                    System.setProperty("mail.mime.multipart.allowempty", "true");
+                    InputStream inputStreams = exchange.getIn().getBody(InputStream.class);
+                    MimeBodyPart mimeMessage = new MimeBodyPart(inputStreams);
+                    DataHandler dh = mimeMessage.getDataHandler();
+                    logger.info("File name: " + dh.getName());
+                    Tika tika = new Tika();
+                    String fileType = tika.detect(dh.getName());
+                    logger.info("File type: " + fileType);
```
Comment (fynmanoj) [inline src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:315]: do not use + operator to concatenate logs , use '{}' format

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java [hunk 5]
```diff
@@ -233,19 +267,158 @@ private void setup() {
...
+                return false;
+        }
+        return true;
+    }
+    private boolean verifyRow(String[] row) {
+            for (int i = 1; i < row.length; i++) {
+                row[i] = row[i].trim();
+                if (row[i].equalsIgnoreCase("MSISDN")) {
+                    logger.info("reached here");
+                    if (!row[i+ 1].matches("^[0-9]*$")) {
```
Comment (fynmanoj) [inline src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:354]: `row[i+1]` can cause  array index out of bound

**Review by fynmanoj — CHANGES_REQUESTED** [review]
changes requested in code

**Review by fynmanoj — CHANGES_REQUESTED** [review]
changes requested in code

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -125,4 +115,9 @@ callback-phases:
...
     - 60
     - 80
-    - 100
\ No newline at end of file
+    - 100
+
+
+csv:
+  columnNames: "id,request_id,payment_mode,payer_identifier_type,payer_identifier,payee_identifier_type,payee_identifier,amount,currency,note"
+  size : 50000 # in bytes
```
[the remark this answers is not in the document]
Comment (ankita10r) [reply src/main/resources/application.yaml:123]: making it 1mb

File: src/main/resources/application-paymentmode.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+payment-mode:
```
[the remark this answers is not in the document]
Comment (ankita10r) [reply src/main/resources/application-paymentmode.yaml:1]: this is how existing functionality was working, should we group it as bulk etc now

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java (hunk 2 above)
Comment (ankita10r) [reply src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:398]: this was a dev logger that i can remove now

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java [hunk 8]
```diff
@@ -233,19 +267,158 @@ private void setup() {
...
 
-
                 });
 
+        from("direct:validateFileSyncResponse")
+                .id("direct:validateFileSyncResponse")
+                .log("Starting route direct:validateFileSyncResponse")
+                .process(exchange -> {
+                    String fileName = System.currentTimeMillis() + "_" +  exchange.getIn().getHeader("fileName", String.class);
+                    String csvData = exchange.getIn().getBody(String.class);
```
[the remark this answers is not in the document]
Comment (ankita10r) [reply src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:271]: These handle values differently so not sure replacing them will not cause regression.

File: src/main/resources/application-paymentmode.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (fynmanoj) [reply src/main/resources/application-paymentmode.yaml:1]: no

File: src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java [hunk 9]
```diff
@@ -233,19 +256,155 @@ private void setup() {
...
+                return false;
+        }
+        return true;
+    }
+    private boolean verifyRow(String[] row) {
+            for (int i = 1; i < row.length; i++) {
+                row[i] = row[i].trim();
+                if (row[i].equalsIgnoreCase("MSISDN")) {
+                    int j = row[i].indexOf("MSISDN");
+                    if (!(j == row.length) && !row[j+ 1].matches("^[0-9]*$")) {
```
Comment (fynmanoj) [inline src/main/java/org/mifos/processor/bulk/camel/routes/ProcessorStartRoute.java:342]: this will cause array index out of bound

**Pull request merged**